### PR TITLE
chore: add FetchEsque type for fetch compatibility

### LINKED ISSUE #2728: feat: type improvements: `typeof fetch` -> `FetchEsque`
> Thanks! LGTM. Just a few linting issues, not sure if you care about those. But solution wise, I think it's a good direction!
> 
> One other thing I'd like to see is fetch-compatible interface being tRPC-driven, rather than `typeof fetch`. Maybe I can take a stub at this?
> 
> _Originally posted by @moltar in https://github.com/trpc/trpc/pull/2723#pullrequestreview-1110792706_

- [ ] `typeof fetch` -> `FetchEsque`
- [ ] `typeof AbortController` -> `AbortControllerEsque`
- [ ] Tests for different fetch-implementations 
- [ ] `Response` --> ??? (TBD)

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Closes #2728

## 🎯 Changes

Adds a `FetchEsque` type that acts as a rough subset of the native `typeof fetch`. Each native interface referenced by it is also given an `interface *Esque` that contains only the members that I could find in tRPC source (or thought were reasonable to include otherwise).

This work was actually a bit trickier than it looks at first..

* Third party packages such as `undici` and `node-fetch` that write their _own_ `fetch`/`Request`/`Response`/etc. types are inherently not compatible with native types or `FetchEsque`. For example, the `forEach` on `Headers` in `node-fetch` doesn't have a third `parent: Headers` parameter in its `callback` parameter.
* I couldn't find a good way to import the Deno fetch specifically, to test it. Am I missing something?

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made. -- not sure where this should go, if anywhere?
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Review by KATT — COMMENTED** [review]
Some half-baked thoughts that isn't really a review yet

**Review by sachinraja — COMMENTED** [review]
Looks good!

File: packages/client/src/index.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export * from './createTRPCClient';
 export * from './createTRPCClientProxy';
 export * from './getFetch';
+export * from './internals/types';
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/client/src/index.ts]: Yeah the only tricky part might be if some external consumer wants to refer to the `FetchEsque` type. It refers to a bunch of other internal types. But there's nothing stopping them from writing a similar type / copy & pasting it in the interim and filing an issue.

File: packages/client/src/internals/types.test.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    ).returns.toEqualTypeOf<Promise<ResponseEsque>>();
+
+    getFetch({} as unknown as typeof fetch);
+  });
+
+  test('NativeFetchEsque', () => {
+    getFetch({} as unknown as typeof import('isomorphic-fetch'));
+    getFetch({} as unknown as typeof import('node-fetch')['default']);
+    getFetch({} as unknown as typeof import('undici')['fetch']);
+  });
```
Comment (juliusmarminge) [inline packages/client/src/internals/types.test.ts:48]: why not

```ts
// top level
import isoFetch from "isomorphic-fetch";
...

  test('NativeFetchEsque', () => {
    getFetch(isoFetch);
    getFetch(nodeFetch);
    getFetch(undici);
  });
```

Comment (JoshuaKGoldberg) [reply packages/client/src/internals/types.test.ts]: 😂 that makes much more sense. Thanks!

**Review by juliusmarminge — DISMISSED** [review]

File: packages/client/src/internals/types.test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  test('NativeFetchEsque', () => {
+    getFetch(isomorphicFetch);
+    getFetch(nodeFetch);
+
+    // Passing in undiciFetch directly in Node v18.7.0 gives:
+    // ReferenceError: TextEncoder is not defined
+    // 🤷
+    getFetch({} as unknown as typeof undiciFetch);
```
Comment (JoshuaKGoldberg) [inline packages/client/src/internals/types.test.ts:54]: This one really stumped me 😕 

https://github.com/trpc/trpc/actions/runs/3873482321/jobs/6603548412

```plaintext
@trpc/tests:test-ci: FAIL client/src/internals/types.test.ts
@trpc/tests:test-ci:   ● Test suite failed to run
@trpc/tests:test-ci: 
@trpc/tests:test-ci:     ReferenceError: TextEncoder is not defined
@trpc/tests:test-ci: 
@trpc/tests:test-ci:       2 | import isomorphicFetch from 'isomorphic-fetch';
@trpc/tests:test-ci:       3 | import nodeFetch from 'node-fetch';
@trpc/tests:test-ci:     > 4 | import { fetch as undiciFetch } from 'undici';
@trpc/tests:test-ci:         | ^
@trpc/tests:test-ci:       5 | import { getFetch } from '../getFetch';
@trpc/tests:test-ci:       6 | import { getAbortController } from './fetchHelpers';
@trpc/tests:test-ci:       7 | import {
```

Node version is 18.7.0, which should support `TextEncoder`. I can repro the issue locally.

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Pull request merged**